src/Common/src/Interop/Windows/kernel32/Interop.CreateFile.cs

```diff
@@ -36,7 +36,9 @@ internal partial class Kernel32
             lpFileName = PathInternal.EnsureExtendedPrefixOverMaxPath(lpFileName);
             fixed (SECURITY_ATTRIBUTES* sa = &securityAttrs)
             {
-                return CreateFilePrivate(lpFileName, dwDesiredAccess, dwShareMode, sa, dwCreationDisposition, dwFlagsAndAttributes, hTemplateFile);
+                return new SafeFileHandle(
```

Comment:
This will leak the filehandle if allocation of the SafeFileHandle fails with OOM.

The built-in PInvoke marshaling deals with this problem by allocating the SafeFileHandle before the call.

Reply:
Good to know, but I don't think it matters in this case. If we're OOM the process is pretty much toast- when it goes down the handle will be released. I think that's reasonable to live with given the perf impact (> 5%).

Reply:
I do not think we should be giving up on robustness like this. We should be cleaning up on errors, even when they are rare.

>  I think that's reasonable to live with given the perf impact (> 5%).

Where do you see this perf impact?

Reply:
> We should be cleaning up on errors, even when they are rare.

This one does clean itself, albeit not immediately. If you feel strongly about it I'm happy to put a try catch around this.

> Where do you see this perf impact?

I didn't look too deeply into it as I didn't see anything super obvious that could be addressed. I got consistent numbers going over large sets of files and chalked it up to the extra allocations and critical finalizer overhead. A few hundred or thousand of these definitely have a big impact.

Reply:
> This one does clean itself

How? If the SafeHandle allocation fails, the handle is leaked.

Reply:
> If we're OOM the process is pretty much toast

OOMs can be a temporary condition, and OOM exceptions are catchable.

Reply:
> How? If the SafeHandle allocation fails, the handle is leaked.

As I stated, with process shutdown.

> OOMs can be a temporary condition, and OOM exceptions are catchable.

I didn't understand that was possible. We always treated them as beyond recovery.

Reply:
> As I stated, with process shutdown.

By that definition, there's never a leak of pretty much anything, as almost every kind of resource is released at process shutdown. That is not a definition we should follow.

Reply:
> That is not a definition we should follow.

Clearly not in the normal case, but you did ask. As I said, I didn't realize that OOM for creating a small object was something that was expected to be recoverable. Given that mistaken assumption hopefully you can see where I was coming from. Not everything is actually released, this is, and I assumed that the process was headed down.